### ISSUE TITLE
How to add "include" section of tsconfig to tsProject?

### ISSUE THREAD

**Issue by IgorBabkin** [ivogabe/gulp-typescript#320]
How to add "include" section of tsconfig to tsProject?

**Comment by ivogabe** [comment]
That is not yet supported by TypeScript, so it's not possible to use that at the moment.

**Comment by IgorBabkin** [comment]
TypeScript supports it. Because I use it already by tsc.

**Comment by ivogabe** [comment]
I did a quick test, and I'm pretty sure it isn't implemented yet. Do you have any reference that says that it was implemented? You can use `files` or `exclude` in a tsconfig file. If you really need a glob, you can use `gulp.src('...')` instead.

**Comment by IgorBabkin** [comment]
hmm, may be I'm wrong. I'm using

{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "inlineSourceMap": true
  },
  "include": [
    "typings/tsd.d.ts"
  ],
  "exclude": [
    "node_modules"
  ]
}

Okay, how to give typescript info about "typings/tsd.d.ts"?

2016-04-07 19:40 GMT+03:00 Ivo Gabe de Wolff <notifications@github.com>:

**Comment by ivogabe** [comment]
All TypeScript files that are in the (subdirectories of) the root directory (directory containing the `tsconfig` file) are passed to TypeScript, unless they were referenced in the `exclude` section. Thus, `typings/tsd.d.ts` will be passed to TypeScript.

**Comment by IgorBabkin** [comment]
var tsProject = ts.createProject(Object.assign({
        declaration: true,
        noExternalResolve: true
    }, tsconfig.compilerOptions));

    var tsStream = gulp.src(['src/*.ts', 'typings/tsd.d.ts'])
        .pipe(ts(tsProject));

    return merge([
        tsStream.dts.pipe(gulp.dest('dist')),
        tsStream.js.pipe(gulp.dest('dist'))
    ])

This is my task and it gives me an error
<img width="461" alt="2016-04-07 10 53 49 pm" src="https://cloud.githubusercontent.com/assets/1428790/14364903/aed42762-fd13-11e5-9876-36ba86bc471e.png">

**Comment by ivogabe** [comment]
For file resolution based on a tsconfig file, you must use `tsProject = ts.createProject('tsconfig.json');` and `tsProject.src()` instead of `gulp.src(...)`. Your config would also work if you removed `noExternalResolve: true`.

**Comment by IgorBabkin** [comment]
But `tsc` copy the whole file/directory structure. I've removed `noExternalResolve: true`. But has the same problem.

**Comment by ivogabe** [comment]
Can you try this:
```javascript
var tsProject = ts.createProject('tsconfig.json', { declaration: true });

var tsStream = tsProject.src()
    .pipe(ts(tsProject));

return merge([
    tsStream.dts.pipe(gulp.dest('dist')),
    tsStream.js.pipe(gulp.dest('dist'))
])
```

**Comment by ivogabe** [comment]
Has this been solved?

**Comment by pbalaga** [comment]
Typescript already does support `include` but only @2.0. See [release notes](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#glob-support-in-tsconfigjson).

`gulp-typescript` still uses v.1.8 which is perfectly reasonable. The problem is that even if you customize the typescript version as below:

`ts.createProject('tsconfig.json', {typescript: require('typescript')});`

the new options are not passed to the compiler. By "new" I mean those which gain support in 2.0, specifically, `include`. Since, I'm using a project file, I would expect `gulp-typescript` to pass all of it to `tsc`, even if some parameters are not recognized or at least tell me that it is skips some of them. It sort of silences out the problem.

Obviously, executing `$(npm bin)/tsc -p tsconfig.json` directly gives expected results.

Is there some work being done regarding transition to ts 2.0? Is it easy to add new parameters?

**Comment by pbalaga** [comment]
I've wired up a quick implementation: see my [fork](https://github.com/pbalaga/gulp-typescript/commit/8239c8d28ec22fb2b7dc2b34dde9c60a59bd4d73).

Required changes are mostly in the project.ts file in how it handles the tsconfig.json. I didn't install TS 2.0 as a submodule and did not write tests though.

Let me know if this is useful.

**Comment by pbalaga** [comment]
As a side note: because of the way how project config options are handled in `gulp-typescript`, the `include` option can be supported without TypeScript 2.0 compiler. Files are searched for before they are passed to the compiler.

**Comment by ivogabe** [comment]
@pbalaga `tsProject.src()` does not use the TypeScript api indeed. It's no problem for me that it `includes` will be supported on TS1.8 too.
Your changes in your fork look good; can you create a pull request? A few comments: `let files = []` should get a type annotation (since it is now typed as `any[]`); you wrote something like `if (!includes) {} if (includes)`, you should replace that with an if-else; and you changed some formatting, can you revert those changes?

**Comment by ivogabe** [comment]
Fixed by @pbalaga in #402 🎉
